kubernetes/deployments_kubeclient.go

```diff
@@ -0,0 +1,1339 @@
+	GetSpace(spaceName string) (*app.SimpleSpace, error)
+	GetApplication(spaceName string, appName string) (*app.SimpleApp, error)
+	GetDeployment(spaceName string, appName string, envName string) (*app.SimpleDeployment, error)
+	ScaleDeployment(spaceName string, appName string, envName string, deployNumber int) (*int, error)
+	GetDeploymentStats(spaceName string, appName string, envName string,
+		startTime time.Time) (*app.SimpleDeploymentStats, error)
+	GetDeploymentStatSeries(spaceName string, appName string, envName string, startTime time.Time,
+		endTime time.Time, limit int) (*app.SimpleDeploymentStatSeries, error)
+	GetEnvironments() ([]*app.SimpleEnvironment, error)
+	GetEnvironment(envName string) (*app.SimpleEnvironment, error)
+	GetPodsInNamespace(nameSpace string, appName string) ([]v1.Pod, error)
+}
+
+type kubeClient struct {
+	config *KubeClientConfig
+	envMap map[string]string
+	KubeRESTAPI
+	MetricsInterface
+	BuildConfigInterface
+}
+
+// KubeRESTAPI collects methods that call out to the Kubernetes API server over the network
+type KubeRESTAPI interface {
+	corev1.CoreV1Interface
+}
+
+type deployment struct {
+	dcUID      types.UID
+	appVersion string
+	current    *v1.ReplicationController
+}
+
+type route struct {
+	host string
+	path string
+	tls  bool
+	// Scoring criteria below
+	hasAdmitted          bool
+	hasAlternateBackends bool
+	isCustomHost         bool
+}
+
+// Receiver for default implementation of KubeRESTAPIGetter and MetricsGetter
+type defaultGetter struct{}
+
+// NewKubeClient creates a KubeClientInterface given a configuration
+func NewKubeClient(config *KubeClientConfig) (KubeClientInterface, error) {
+	// Use default implementation if no KubernetesGetter is specified
+	if config.KubeRESTAPIGetter == nil {
+		config.KubeRESTAPIGetter = &defaultGetter{}
+	}
+	kubeAPI, err := config.GetKubeRESTAPI(config)
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+
+	// Use default implementation if no MetricsGetter is specified
+	if config.MetricsGetter == nil {
+		config.MetricsGetter = &defaultGetter{}
+	}
+	// In the absence of a better way to get the user's metrics URL,
+	// substitute "api" with "metrics" in user's cluster URL
+	metricsURL, err := getMetricsURLFromAPIURL(config.ClusterURL)
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+	// Create MetricsClient for talking with Hawkular API
+	metricsConfig := &MetricsClientConfig{
+		MetricsURL:  metricsURL,
+		BearerToken: config.BearerToken,
+	}
+	metrics, err := config.GetMetrics(metricsConfig)
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+
+	kubeClient := &kubeClient{
+		config:               config,
+		KubeRESTAPI:          kubeAPI,
+		MetricsInterface:     metrics,
+		BuildConfigInterface: config.BuildConfigInterface,
+	}
+
+	// Get environments from config map
+	envMap, err := kubeClient.getEnvironmentsFromConfigMap()
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+	kubeClient.envMap = envMap
```

Comment:
As I read it, getEnvironmentsFromConfigMap() needs the partially constructed kubeClient in order to make a request to fill in the envMap.  While it's true it could fill in the kubeClient directly, it doesn't.

Reply:
Yes, although we could change this to a function and pass the information we need as arguments if that is preferred.

Reply:
ok